design/one-pager-stack-ui-metadata.md

```diff
@@ -0,0 +1,155 @@
+# Stack Package UI configuration metadata
+
+* Owner: Steven Rathbauer (@rathpc), Marques Johansson (@displague)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Terms
+
+* **Stack**: A Stack is a Crossplane extension-manager managed package.  These packages may bundle one or many Crossplane applications or infrastructure provider controllers.
+* **Stacks**: Stacks could refer to more than one Crossplane extension-manager packages or the Extension-Manager system as a whole.
+
+See the [Stacks Design Doc](design-doc-stacks.md) for more details about Crossplane Stacks.
+
+## Objective
+
+This document aims to provide details on the necessary metadata that will drive the UI on a configuration page with respect to Crossplane Stacks and the user configurable custom-resource fields exposed with a Stack. This is meant to be a dynamic spec that will not attempt to accomodate all potentional UI elements.  This spec does not intend to be as complete as [XForms](https://en.wikipedia.org/wiki/XForms) or [HTML Forms](https://www.w3.org/TR/html52/sec-forms.html), for example.
+
+Users may choose to annotate Stack bundled CRD today with UI hints.  This proposal suggests a means for the extension manager to apply CRD annotations at install and upgrade time.  This separation provides a better developer experience preventing the need for modifying escaped, nested, quoted, YAML or JSON annotations within a CRD document while iterating over UI design.
+
+A YAML file named `ui-schema.yaml` is proposed for inclusion in Stacks.  This will standardize the approach Stack creators use to offer UI context to their package and offer a more engaging user experience.
+
+Tools that deliver a user-interface based on this proposal may include:
+
+* CLI help output
+* CLI auto-completion for arguments
+* Advanced validation (beyond what Kubernetes' OAS3 support delivers)
+* ncurses-style installers
+* web-based package managers (out-of-cluster and in-cluster)
+* web-based configuration panels
+
+This spec is not concerned with a specific UI implementation but serves to provide a foundation for supporting Stack developers and users through basic form input.
+
+## Proposed File format
+
+The `ui-schema.yaml` file will conform to a specification that will not be covered formally in this document but will be suggested through examples and supporting text.
```

Comment:
this sentence could use a bit more explanation around why we are not covering a formal specification, nor linking to an external formal specification.  The reason we aren't doing that is because we do not want to be prescriptive or authoritative about the capabilities of any UI that could be driven from this content, we want to be neutral and agnostic about this.

Reply:
That's fair.  My reasoning for adding this was to the avoid creating a schema file because I couldn't find an appropriate validation spec or tool.  I assumed we could reasonably validate on some field names and values without blocking additions.

If the intent is to leave the format wide open without any meaningful validation, that does simplify things, including the implementation:  `yaml` -> `json` -> annotatation.

Reply:
Explained in 418e3ab4041f038ba50eac0477bea30fe1cac494